Adds apple-touch-icon to /bus-arrival/

jglim opened the pull request:
Default busrouter.sg apple-touch-icon added to /bus-arrival/, inserted
inline as it may be overwritten below:

On /bus-arrival/, when a bus is pinned, the default apple-touch-icon is
overridden with a new canvas-generated icon for that specific bus

---

Here's a screenshot of the generated apple-touch-icon ("Clementi")

![4348bfd](https://user-images.githubusercontent.com/1116555/49227167-6debc200-f423-11e8-92e2-8b1cf342b918.png)

I've also uploaded a [screen recording](http://jg.sn.sg/e/4348bfd.mov) of the result
- 00:00 to 00:04 : default behavior with the new apple-touch-icon
- 00:04 to 00:13 : changes to apple-touch-icon after pinning
- 00:13 to 00:17 : launching from the bookmark (no address bar yay!)

cheeaun commented on bus-arrival/index.html:
```diff
@@ -1,6 +1,8 @@
 <meta charset="utf-8">
 <meta name="viewport" content="width=device-width, initial-scale=1.0">
+<meta name="apple-mobile-web-app-capable" content="yes">
+<link rel="apple-touch-icon-precomposed" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAgAAAAIACAMAAADDpiTIAAABwlBMVEX///8ARU3vLC5ciI3i6uv9/v5GeH41bHJ5naFzmZ36+/xymJ3z9vebtrkDR0/vLS8IS1MOT1cESFAeW2KHp6shXmXwNDbC09Xf6Onc5uf4+vo8cXdSgYYUVFsLTVX95ubJ2Nq8ztAlYGf//f381dXq8PC0yctNfYPu8/PR3d9Wg4lKe4BvlpoQUVnvMTPG1df++PjY4+SxxsmWs7YbWWDxQkN8n6T0aGqqwcRZhovw9PVPf4UqZGrxRUfm7e796urM2ttnkJVhjJGQrrL+7+ ...
```
First, can get rid of the `-precomposed` part because it's no longer needed since iOS 7.

Second, can link to the original icon image here `icons/icon-192.png` instead of this data URI?

jglim replied:
:+1:  implemented in https://github.com/cheeaun/busrouter-sg/commit/fd739242ffd5e83af6bd7b393727e9b3f5aac6ff

jglim commented:
Thanks for the quick reply!

- _Since we're already setting a custom icon, why not set a custom app title as well? By doing this \<meta name="apple-mobile-web-app-title" content="XXX"\>_

Sure! I used to save the stop names based on intent (e.g. 196 "Home » Clementi"), though displaying the stop name might be semantically better, so I implemented it as such.

- _Would also be cool if also can generate an icon for a bus stop too, instead of just a bus service number._

Static image (can!) or dynamic content generated from text (trickier, still can..)? Will first listen to and understand your thoughts before implementing.

- _However, what happens when someone "stars" more than one bus service code?_

The last "starred" service code will be the one that's used as the icon :laughing: Will be happy to change that, though I am not sure how to approach it yet.

https://github.com/cheeaun/busrouter-sg/pull/29/commits/fd739242ffd5e83af6bd7b393727e9b3f5aac6ff 

- "apple-touch-icon-precomposed" is now "apple-touch-icon"
- apple-touch-icon links to a png instead of inlining base64 (still works)
- Added apple-mobile-web-app-title that contains the current bus stop's
name

[screen recording](http://jg.sn.sg/e/fd73924.mov)

cheeaun commented:
Hey @jglim sorry for the late reply.

After some thinking, here are my thoughts:

- The icon shows bus stop code instead of bus service number.
- App title will show bus stop name instead. Since it's editable, users can still rename it to whatever they want (Maybe bus service numbers for your use-case).

There can be a lot of permutations for this but let's keep it simple for now 😁

jglim commented:
Hi @cheeaun - no problem :smiley: 
The changes requested are available in 807d5e9

[screen recording, again](http://jg.sn.sg/e/807d5e9.mov)

cheeaun commented on assets/arrival.js:
```diff
@@ -131,6 +132,7 @@ class ArrivalTimes extends Component {
         if (stop) {
           const [lng, lat, name] = stop;
           document.title = `Bus arrival times for ${code + ' - ' + name}`;
+          document.querySelectorAll('[name="apple-mobile-web-app-title"]')[0].setAttribute("content", name);
```
Could use `document.querySelector` instead which returns the first element instead of an array of elements. Also `"content"` should be single-quoted.

cheeaun commented on bus-arrival/index.html:
```diff
@@ -1,7 +1,10 @@
+<meta name="apple-mobile-web-app-capable" content="yes">
+<meta name="apple-mobile-web-app-title" content="Bus arrival times">
 <title>Bus arrival times</title>
+<link rel="apple-touch-icon" href="../icons/icon-192.png">
```
Would be great if can move all these meta-related tags below everything else since they are less important than the `<link>` tags. The root `index.html` is an example of how I order them.

cheeaun commented on utils/setIcon.js:
```diff
@@ -0,0 +1,45 @@
+export default (iconText) => {
+  let iconSize = 512;
+  let iconBackgroundColor = "#F01B48";
+  let iconPrimaryColor = "#FFFFFF";
```
Single quotes :)

cheeaun commented on utils/setIcon.js:
```diff
@@ -0,0 +1,45 @@
+  ctx.fillText(iconText, (c.width / 2), (c.height / 2));
+
+  // replace existing apple-touch-icon
+  document.querySelectorAll('[rel="apple-touch-icon"]')[0].setAttribute("href", c.toDataURL());
```
`querySelector` instead of `querySelectorAll`, and single quotes :)

jglim commented:
Thanks for reviewing 807d5e9. I've implemented the requested changes in 333f590. There are no visual changes, so I didn't upload a recording.

cheeaun commented on bus-arrival/index.html:
```diff
@@ -2,13 +2,16 @@
 <meta charset="utf-8">
 <meta name="viewport" content="width=device-width, initial-scale=1.0">
 <title>Bus arrival times</title>
+<link rel="apple-touch-icon" href="../icons/icon-192.png">
```
Can also move this `link` tag below all the other `link` tags since it's less important :)

jglim commented:
Done! Sorry I didn't catch that one 😅

jglim commented:
@cheeaun Thanks for taking time to help this PR through!
